File: App/Resources/AlgorithmDetails/avltreesort/avltreesort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class avltreesort {
  private static class Node {
    int value;
    Node left;
    Node right;
    int balance;

    Node(int value) {
      this.value = value;
    }
  }

  private static class AddResult {
    Node node;
    boolean heightChanged;

    AddResult(Node node, boolean heightChanged) {
      this.node = node;
      this.heightChanged = heightChanged;
    }
  }

  private static Node singleRotateRight(Node node) {
    Node b = node.left;
    node.left = b.right;
    b.right = node;
    node.balance = 0;
    b.balance = 0;
    return b;
  }

  private static Node singleRotateLeft(Node node) {
    Node b = node.right;
    node.right = b.left;
    b.left = node;
    node.balance = 0;
    b.balance = 0;
    return b;
  }

  private static Node doubleRotateRight(Node node) {
    int oldBBalance = node.left.right.balance;
    node.left = singleRotateLeft(node.left);
    Node b = singleRotateRight(node);
    if (oldBBalance == -1) {
      b.right.balance = 1;
    }
    if (oldBBalance == 1) {
      b.left.balance = -1;
    }
    return b;
  }

  private static Node doubleRotateLeft(Node node) {
    int oldBBalance = node.right.left.balance;
    node.right = singleRotateRight(node.right);
    Node b = singleRotateLeft(node);
    if (oldBBalance == -1) {
      b.right.balance = 1;
    }
    if (oldBBalance == 1) {
      b.left.balance = -1;
    }
    return b;
  }

  private static AddResult heightChangeLeft(Node node) {
    if (node.balance != -1) {
      node.balance -= 1;
      return new AddResult(node, node.balance == -1);
    }
    if (node.left.balance == -1) {
      return new AddResult(singleRotateRight(node), false);
    }
    return new AddResult(doubleRotateRight(node), false);
  }

  private static AddResult heightChangeRight(Node node) {
    if (node.balance != 1) {
      node.balance += 1;
      return new AddResult(node, node.balance == 1);
    }
    if (node.right.balance == 1) {
      return new AddResult(singleRotateLeft(node), false);
    }
    return new AddResult(doubleRotateLeft(node), false);
  }

  private static AddResult add(Node node, int value) {
    if (node == null) {
      return new AddResult(new Node(value), true);
    }
    if (value < node.value) {
      AddResult result = add(node.left, value);
      node.left = result.node;
      if (result.heightChanged) {
        return heightChangeLeft(node);
      }
      return new AddResult(node, false);
    } else {
      AddResult result = add(node.right, value);
      node.right = result.node;
      if (result.heightChanged) {
        return heightChangeRight(node);
      }
      return new AddResult(node, false);
    }
  }

  private static void traverse(Node node, List<Integer> result) {
    if (node == null) {
      return;
    }
    traverse(node.left, result);
    result.add(node.value);
    traverse(node.right, result);
  }

  public static void sort(int[] arr) {
    Node root = null;
    for (int value : arr) {
      root = add(root, value).node;
    }

    List<Integer> result = new ArrayList<>();
    traverse(root, result);

    for (int i = 0; i < arr.length; i++) {
      arr[i] = result.get(i);
    }
  }

  public static void main(String[] args) {
    int[] array = new int[] {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}

File: App/Resources/AlgorithmDetails/avltreesort/avltreesort.rb
class Node
  attr_accessor :value, :left, :right, :balance

  def initialize(value)
    @value = value
    @left = nil
    @right = nil
    @balance = 0
  end
end

AddResult = Struct.new(:node, :height_changed)

def single_rotate_right(node)
  b = node.left
  node.left = b.right
  b.right = node
  node.balance = 0
  b.balance = 0
  b
end

def single_rotate_left(node)
  b = node.right
  node.right = b.left
  b.left = node
  node.balance = 0
  b.balance = 0
  b
end

def double_rotate_right(node)
  old_b_balance = node.left.right.balance
  node.left = single_rotate_left(node.left)
  b = single_rotate_right(node)
  b.right.balance = 1 if old_b_balance == -1
  b.left.balance = -1 if old_b_balance == 1
  b
end

def double_rotate_left(node)
  old_b_balance = node.right.left.balance
  node.right = single_rotate_right(node.right)
  b = single_rotate_left(node)
  b.right.balance = 1 if old_b_balance == -1
  b.left.balance = -1 if old_b_balance == 1
  b
end

def height_change_left(node)
  if node.balance != -1
    node.balance -= 1
    return AddResult.new(node, node.balance == -1)
  end
  return AddResult.new(single_rotate_right(node), false) if node.left.balance == -1

  AddResult.new(double_rotate_right(node), false)
end

def height_change_right(node)
  if node.balance != 1
    node.balance += 1
    return AddResult.new(node, node.balance == 1)
  end
  return AddResult.new(single_rotate_left(node), false) if node.right.balance == 1

  AddResult.new(double_rotate_left(node), false)
end

def add(node, value)
  return AddResult.new(Node.new(value), true) if node.nil?

  if value < node.value
    result = add(node.left, value)
    node.left = result.node
    return height_change_left(node) if result.height_changed
  else
    result = add(node.right, value)
    node.right = result.node
    return height_change_right(node) if result.height_changed
  end

  AddResult.new(node, false)
end

def traverse(node, result)
  return if node.nil?

  traverse(node.left, result)
  result << node.value
  traverse(node.right, result)
end

def sort(arr)
  root = nil
  arr.each { |value| root = add(root, value).node }

  result = []
  traverse(root, result)

  (0...arr.length).each { |i| arr[i] = result[i] }
  arr
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array
